x-pack/plugins/canvas/canvas_plugin_src/functions/types/functions.ts

```diff
@@ -0,0 +1,112 @@
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+import { ArgumentType, TypeToCanvasArgument } from './arguments';
+import { functions as commonFunctions } from '../../functions/common';
+import { functions as browserFunctions } from '../../functions/browser';
+import { functions as serverFunctions } from '../../functions/server';
+
+/**
+ * A Function type which represents a Function in Canvas.  This type assumes
+ * any Context can be provided and used by the Function implementation.
+ */
+export interface FunctionSpec<Name, Arguments, Return> {
+  /** Arguments for the Function */
+  args: { [key in keyof Arguments]: ArgumentType<Arguments[key]> };
+  aliases?: string[];
+  /** Help text displayed in the Expression editor */
+  help: string;
+  /** The name of the Function */
+  name: Name;
+  /** The type of the Function */
+  type?: CanvasFunctionType | Name;
+  /** The implementation of the Function */
+  fn(context: any, args: Arguments, handlers: FunctionHandlers): Return;
+}
+
+/**
+ * A Function type which restricts the incoming Context to a specific type.
+ */
+export interface ContextFunctionSpec<Name, Context, Arguments, Return>
+  extends FunctionSpec<Name, Arguments, Return> {
+  /** The incoming Context provided to the Function; the information piped in. */
+  context?: {
+    types: Array<TypeToCanvasArgument<Context>>;
+  };
+  /** The implementation of the Function */
+  fn(context: Context, args: Arguments, handlers: FunctionHandlers): Return;
+}
+
+/**
+ * A Function type which restricts the incoming Context specifically to `null`.
+ */
+export interface NullContextFunctionSpec<Name, Arguments, Return>
+  extends ContextFunctionSpec<Name, null, Arguments, Return> {
+  /** The incoming Context provided to the Function; the information piped in. */
+  context?: {
+    types: ['null'];
+  };
+  /** The implementation of the Function */
+  fn(context: null, args: Arguments, handlers: FunctionHandlers): Return;
+}
+
+/**
+ * A type which infers all of the Function names.
+ */
+// prettier-ignore
+export type AvailableFunctions<FnFactory> = 
+  FnFactory extends ContextFunctionFactory<infer Name, infer Context, infer Arguments, infer Return> ?
+    { name: Name, context: Context, arguments: Arguments, return: Return } :
+  FnFactory extends NullContextFunctionFactory<infer Name, infer Arguments, infer Return> ?
+    { name: Name, arguments: Arguments, return: Return } :
+  FnFactory extends FunctionFactory<infer Name, infer Arguments, infer Return> ?
+    { name: Name, arguments: Arguments, return: Return } :
+    never;
+
+/**
+ * A type containing all of the Function names available to Canvas, formally exported.
+ */
+// prettier-ignore
+export type AvailableFunctionNames = 
+  AvailableFunctions<typeof commonFunctions[number]>['name'] &
```

Comment:
First of all, incredible work, learning so much just by looking through this! How can `AvailableFunctionNames` be used? It's not being referred to now, does it have an impact without use, or if not, what can in theory be hooked up to it?

Reply:
This is going to be used for i18n, for starters.  You'll see it's usefulness in my next diff.

Reply:
Here's a practical example.  In this case, the `AvailableFunctions` type can be used to enforce the structure of the `FunctionHelpDict` type, throwing a TS error if any of the functions are missing.

```js
/** This type uses the collection of available functions, (which are all
 * `FunctionFactory`), to derive the name and argument types from them.  This
 * allows for validation that 1/ every function available is listed, and 2/ every
 * function argument has a help string present.
 */
type FunctionHelp<T> = T extends FunctionFactory<
  infer Name,
  infer Arguments,
  infer Context,
  infer Errors
>
  ? {
      [key in Name]: {
        help: string;
        args: { [key in keyof Arguments]: string };
      }
    }
  : never;

/**
 * This type represents an exhaustive dictionary of Function help strings,
 * organized by Function and then Function Argument.
 *
 * This type indexes the existing function factories, reverses the union to an
 * intersection, and produces the dictionary of strings.
 */
type FunctionHelpDict = UnionToIntersection<FunctionHelp<typeof functions[number]>>;

/**
 * Help text for Canvas Functions should be properly localized. This function will
 * return a dictionary of help strings, organized by Canvas Function specification
 * and then by available arguments.
 */
export const getFunctionHelp = (): FunctionHelpDict => {
  return {
    alterColumn: {
      help: i18n.translate('xpack.canvas.functions.alterColumnHelpText', {
        defaultMessage:
          'Converts between core types, eg string, number, null, boolean, date and rename columns',
      }),
      args: {
        column: i18n.translate('xpack.canvas.functions.alterColumn.args.columnHelpText', {
          defaultMessage: 'The name of the column to alter',
        }),
        type: i18n.translate('xpack.canvas.functions.alterColumn.args.typeHelpText', {
          defaultMessage: 'The type to convert the column to. Leave blank to not change type',
        }),
        name: i18n.translate('xpack.canvas.functions.alterColumn.args.nameHelpText', {
          defaultMessage: 'The resultant column name. Leave blank to not rename',
        }),
      },
    },
```